reference-implementation/to-upstream-wpts/writable-streams/aborting.js

```diff
@@ -43,5 +41,256 @@ promise_test(t => {
+}, 'Fulfillment value of ws.abort() call must be undefined even if the underlying sink returns a non-undefined value');
+
+promise_test(() => {
+  const errorInSinkAbort = new Error('Sorry, it just wasn\'t meant to be.');
+  const ws = new WritableStream({
+    abort() {
+      throw errorInSinkAbort;
+    }
+  });
+
+  const writer = ws.getWriter();
+
+  const abortPromise = writer.abort(undefined);
+  return abortPromise.then(
+    () => {
+      throw new Error('abortPromise is fulfilled unexpectedly');
+    },
+    r => {
+      assert_equals(r, errorInSinkAbort, 'rejection reason of abortPromise must be errorInSinkAbort');
+    }
+  );
+}, 'WritableStream if sink\'s abort throws, the promise returned by writer.abort() rejects');
+
+promise_test(() => {
+  const errorInSinkAbort = new Error('Sorry, it just wasn\'t meant to be.');
+  const ws = new WritableStream({
+    abort() {
+      throw errorInSinkAbort;
+    }
+  });
+
+  const abortPromise = ws.abort(undefined);
+  return abortPromise.then(
+    () => {
+      throw new Error('abortPromise is fulfilled unexpectedly');
+    },
+    r => {
+      assert_equals(r, errorInSinkAbort, 'rejection reason of abortPromise must be errorInSinkAbort');
+    }
+  );
+}, 'WritableStream if sink\'s abort throws, the promise returned by ws.abort() rejects');
+
+test(() => {
+  let recordedReason;
+  const ws = new WritableStream({
+    abort(reason) {
+      recordedReason = reason;
+    }
+  });
+
+  const writer = ws.getWriter();
+
+  const passedReason = new Error('Sorry, it just wasn\'t meant to be.');
+  writer.abort(passedReason);
+
+  assert_equals(recordedReason, passedReason);
+}, 'Aborting a WritableStream passes through the given reason');
+
+promise_test(() => {
+  const ws = new WritableStream();
+  const writer = ws.getWriter();
+
+  const passedReason = new Error('Sorry, it just wasn\'t meant to be.');
+  writer.abort(passedReason);
+
+  return Promise.all([
+    writer.write().then(
+      () => { throw new Error('writing should not succeed'); },
+      r => assert_equals(r.constructor.toString(), TypeError.toString(), 'writing should reject with the given reason')
+    ),
+    writer.close().then(
+      () => { throw new Error('closing should not succeed'); },
+      r => assert_equals(r.constructor.toString(), TypeError.toString(), 'closing should reject with the given reason')
+    ),
+    writer.abort().then(
+      () => { throw new Error('aborting a second time should not succeed'); },
+      r => assert_equals(r.constructor.toString(), TypeError.toString(), 'aborting a second time should reject ' +
+        'with the given reason')
+    ),
+    writer.closed.then(
+      () => { throw new Error('closed fulfilled unexpectedly'); },
+      r => assert_equals(r.constructor.toString(), TypeError.toString(), 'closed should reject with a TypeError')
+    )
+  ]);
+}, 'Aborting a WritableStream puts it in an errored state, with stored error equal to the abort reason');
+
+test(() => {
```

Comment:
I'm unsure where to import `assert_promise_rejects` from. This function seems to be duplicated inside the `web-platform-tests` repo. I tried including one of the existing functions, but get an URL error by testharness